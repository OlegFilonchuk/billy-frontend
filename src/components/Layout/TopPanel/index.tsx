import React, { FC } from 'react';
import { BiMenuAltLeft, FiArrowLeft, HiDotsVertical } from 'react-icons/all';
import { useTranslation } from 'react-i18next';
import { useLocation, useHistory } from 'react-router-dom';
import { IconContext } from 'react-icons';
import Button from '../../common/Button';
import logo from './billy_logo.png';
import { ROUTES } from '../../../constants';

type Props = {
  onRangesToggle: () => void;
  onMenuToggle: () => void;
};

const TopPanel: FC<Props> = ({ onRangesToggle, onMenuToggle }) => {
  const { t } = useTranslation();

  const location = useLocation();
  const history = useHistory();

  const handleBack = () => {
    history.goBack();
  };

  return (
    <div className="bg-main h-16 w-full text-textLight flex flex-row flex-nowrap items-center justify-between p-2 shadow-sm fixed z-50">
      <IconContext.Provider value={{ size: '24' }}>
        {location.pathname === ROUTES.root && (
          <>
            <Button onClick={onRangesToggle}>
              <BiMenuAltLeft />
            </Button>

            <div className="flex-1 h-full">
              <img src={logo} alt="Billy" className="h-6 object-contain" />
            </div>

            <Button onClick={onMenuToggle}>
              <HiDotsVertical />
            </Button>
          </>
        )}

        {(location.pathname === ROUTES.expense ||
          location.pathname === ROUTES.income) && (
          <>
            <Button onClick={handleBack}>
              <FiArrowLeft />
            </Button>

            <div className="flex-1 h-full flex items-center px-4">
              {t(
                location.pathname === ROUTES.expense
                  ? 'New Expense'
                  : 'New Income',
              )}
            </div>

            <div />
          </>
        )}
      </IconContext.Provider>
    </div>
  );
};

export default TopPanel;
