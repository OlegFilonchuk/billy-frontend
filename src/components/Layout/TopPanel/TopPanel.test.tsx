import { fireEvent, RenderResult } from '@testing-library/react';
import React from 'react';
import TopPanel from './TopPanel';
import renderWithI18n from '../../../utils/testUtills';

jest.mock('react-router-dom', () => ({
  useLocation: () => ({
    pathname: '/',
  }),
  useHistory: () => ({
    goBack: jest.fn(),
  }),
}));

describe('TopPanel', () => {
  let ui: RenderResult;
  const onRangesToggle = jest.fn();
  const onMenuToggle = jest.fn();

  describe('with "withBack" prop set to true or unset', () => {
    beforeEach(() => {
      ui = renderWithI18n(
        <TopPanel
          onMenuToggle={onMenuToggle}
          onRangesToggle={onRangesToggle}
        />,
      );
    });

    it('should render properly', () => {
      expect(ui.container).toMatchSnapshot();
    });

    it('should contain back button by default', () => {
      const backButton = ui.getByTitle('Go back');
      expect(backButton).toBeInTheDocument();
    });

    it('should contain "Select Date Range" button if pathname is "/"', () => {
      const rangeButton = ui.getByTitle('Select Date Range');
      expect(rangeButton).toBeInTheDocument();
    });

    it('should contain "Menu" button if pathname is "/"', () => {
      const menuButton = ui.getByTitle('Menu');
      expect(menuButton).toBeInTheDocument();
    });

    it('should contain logo in pathname is "/"', () => {
      const logo = ui.getByAltText('Billy');
      expect(logo).toBeInTheDocument();
    });

    it('should trigger callbacks on buttons click', () => {
      expect(onRangesToggle).not.toHaveBeenCalled();
      expect(onMenuToggle).not.toHaveBeenCalled();

      const rangeButton = ui.getByTitle('Select Date Range');
      const menuButton = ui.getByTitle('Menu');

      fireEvent.click(rangeButton);
      fireEvent.click(menuButton);

      expect(onRangesToggle).toHaveBeenCalledTimes(1);
      expect(onMenuToggle).toHaveBeenCalledTimes(1);
    });
  });

  describe('with "withBack" prop set to false', () => {
    beforeEach(() => {
      ui = renderWithI18n(
        <TopPanel
          onMenuToggle={onMenuToggle}
          onRangesToggle={onRangesToggle}
          withBack={false}
        />,
      );
    });

    it('should not contain back button', () => {
      const backButton = ui.queryByTitle('Go back');
      expect(backButton).not.toBeInTheDocument();
    });
  });
});
