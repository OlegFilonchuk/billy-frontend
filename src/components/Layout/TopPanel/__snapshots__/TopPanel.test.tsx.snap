// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopPanel with "withBack" prop set to true or unset should render properly 1`] = `
<div>
  <div
    class="bg-main h-16 w-full text-textLight flex flex-row flex-nowrap items-center justify-between p-2 shadow-sm fixed z-50"
  >
    <button
      class="transition duration-150 hover:shadow-lg focus:shadow-lg focus:ring ring-accent disabled:opacity-50 p-2 rounded inline-flex justify-center items-center    "
      title="Go back"
      type="button"
    >
      <svg
        fill="currentColor"
        height="24"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M244 400L100 256l144-144M120 256h292"
          fill="none"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="48"
        />
      </svg>
    </button>
    <button
      class="transition duration-150 hover:shadow-lg focus:shadow-lg focus:ring ring-accent disabled:opacity-50 p-2 rounded inline-flex justify-center items-center    "
      title="Select Date Range"
      type="button"
    >
      <svg
        fill="currentColor"
        height="24"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M4 11H16V13H4zM4 6H20V8H4zM4 18L11 18 11.235 18 11.235 16 11 16 4 16z"
        />
      </svg>
    </button>
    <div
      class="flex-1 h-full"
    >
      <img
        alt="Billy"
        class="h-6 object-contain"
        src="logo_white.png"
      />
      <div
        class="text-sm px-1"
      >
        All Accounts
      </div>
    </div>
    <button
      class="transition duration-150 hover:shadow-lg focus:shadow-lg focus:ring ring-accent disabled:opacity-50 p-2 rounded inline-flex justify-center items-center    "
      title="Menu"
      type="button"
    >
      <svg
        fill="currentColor"
        height="24"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 20 20"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10 6a2 2 0 110-4 2 2 0 010 4zM10 12a2 2 0 110-4 2 2 0 010 4zM10 18a2 2 0 110-4 2 2 0 010 4z"
        />
      </svg>
    </button>
  </div>
</div>
`;
